### DISCUSSION TITLE
[WASM] OpenFilePickerAsync do nothing on iphone

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#16739] [Q&A]
Hi,
I have a simple WASM app with just a button that call OpenFilePickerAsync to select file. 
The app works as expected on my Android device, and from Desktop, but on iPhone nothing happen (tried on Safari and Edge)... 

Is there some kind of additionnal configuration to do for iPhone device? or is it sopposed to work like on other device?

Thank you for any help.

**Comment by maxkatz6** [comment]
Does CanOpen returns true?
To be clear, it's absolutely possible for some platform to not support specific file pickers. Crossplatform application should always check for CanOpen/CanSave. If these properties values and actual behavior does not match, that's a bug.

**Reply by maxkatz6** [reply]
And specifically, about Safari - it doesn't support openFilePicker https://developer.mozilla.org/en-US/docs/Web/API/Window/showOpenFilePicker
We use a third party polyfill (ponyfill?), but its capabilities are also limited https://github.com/jimmywarting/native-file-system-adapter/blob/master/src/showOpenFilePicker.js

**Reply by dbriard** [reply]
@maxkatz6 I believed that OpenFilePicker was okay for all browsers, so thx, I will always add that test in all my code.

But in this specific case, I actually tested and yes, CanOpen is true on Safari iPhone...

Thx for the link but I do not understand the table, or may be it is obsolete?
![image](https://github.com/user-attachments/assets/f411eb35-70a1-4d36-a4d0-c4823720f1b1)
Because I am using OpenFilePickerAsync on Firefox without any trouble.
Same for my Android browsers, everything work fine (see capture below).

When you say "We use a third party polyfill" you mean in Avalonia (implementation of OpenFilePicker), or in your own programs? **Is there an option to activate to use that alternate file picker?** or do I have to do js interop myself?

Here is the view I have on Android chrome, it look native to me.
![image](https://github.com/user-attachments/assets/bea235e9-6a28-4160-b5f7-3216de17fe4c)

But nothing on iPhone :( I have no exception, no error, and nothing on screen.
I'll try to add a WriteLine just before, and just after the call to OpenFilePickerAsync tomorrow.

I am also wondering if there is something to add somewhere in order to ask permission to access to photo library, or camera? on iPhone.

**Reply by maxkatz6** [reply]
> Because I am using OpenFilePickerAsync on Firefox without any trouble.
> When you say "We use a third party polyfill" you mean in Avalonia (implementation of OpenFilePicker), or in your own programs?

No, I mean this specific third party JS library  https://github.com/jimmywarting/native-file-system-adapter

**Reply by dbriard** [reply]
I tried PreferFileDialogPolyfill on the BrowserPlatformOptions but I do not see any difference.

I am not sure what I changed, may be changing the async routed command to just routed command, but I managed to display the FilePicker on iPhone.

The only issue and I don't know if it is a bug in Avalonia or not, is that the first click on the Button do nothing, then after the first click, the FilePicker work as expected. @maxkatz6, do you know if it is a know bug? and if there is a workaround?

Here is the code I am using:

![image](https://github.com/user-attachments/assets/01bf9899-d6c3-455d-95e7-2719a7f6eff9)

This line is never called if I cancel selecting a file: "Message = "After OpenFilePickerAsync";"
Probably there is no cancel event at your disposal?
That mean it always stay in the await? could it be an issue?
Thank you.
